examples/ota-provider-app/linux/UserConsentProvider.h

```diff
@@ -0,0 +1,62 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+#include <ota-provider-common/UserConsentDelegate.h>
+
+class UserConsentProvider : public chip::ota::UserConsentDelegate
+{
+public:
+    void ObtainUserConsentAsync(chip::NodeId nodeId, chip::EndpointId endpoint, uint32_t currentVersion, uint32_t newVersion)
+    {
+        mNodeId         = nodeId;
+        mEndpoint       = endpoint;
+        mCurrentVersion = currentVersion;
+        mNewVersion     = newVersion;
+
+        // This is not an async operation, std::cin blocks the chip thread
+        // TODO: Spawn a new thread for getting user consent
+        chip::DeviceLayer::SystemLayer().StartTimer(chip::System::Clock::Milliseconds32(1 * 1000), &HandleUserConsent, this);
+    }
+
+private:
+    chip::NodeId mNodeId       = chip::kUndefinedNodeId;
+    chip::EndpointId mEndpoint = chip::kInvalidEndpointId;
+    uint32_t mCurrentVersion   = 0;
+    uint32_t mNewVersion       = 0;
+
+    static void HandleUserConsent(chip::System::Layer * systemLayer, void * args)
```

Comment:
This will block testing in the "default" OTA provider used eventually for test harness and CSG.

I strongly suggest having the delegate, but having the default implementation do the following:

- Have the concrete delegate have a setter with whether consent is given or not
- Allow this value to be set at runtime, but start with a default of true

This will allow tooling based around the sample apps to possibly use means other than console interaction to set/clear consent.

The spec is clear that:

> An OTA Provider SHOULD obtain some form of "User Consent" prior to responding with a URI for a Software Image or letting an OTA Requestor proceed with applying a previously downloaded update. In the context of the `OTA` Cluster, "User Consent" SHALL be defined as any signal that an OTA Provider may obtain through its implementation-specific logic, that conveys consent to proceed from a User administratively allowed to give such consent in an informed manner.

The consent operation is not a SHALL.

Also, one of the suggested methods:

> * Relaying of a previously stored consent signal, previously provided by a User at some point in the past. The original capture of the stored consent signal should have been made after having provided sufficient information to the User to understand the consequences of such stored consent. Multiple signals, covering different Nodes, Vendors or Device Classes, may be stored independently to affect a variety of deferred consent policies.

Here, "previously stored consent signal"  may be the fact that in test environments like the sample OTA provider, there is no customer other than the people testing, and the consent being given by means like the command line should be satisfactory.

If an implementer wants to base off the sampel OTA provider, but set the delegate to something else, they can, and they can then have that async operation proceed.